### PR TITLE
TO_CHAR with Date operand to DATE_FORMAT

### PR DESCRIPTION
**Pull request by Katakam-Phani-Mahidhar**

### CODE REVIEW

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java [hunk 1]
```diff
@@ -890,6 +890,7 @@ private SqlNode callToSql(@Nullable RexProgram program, RexCall rex, boolean not
         }
       case PLUS:
       case MINUS:
+      case OTHER_FUNCTION:
```
Comment (saranshmanglani) [inline core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java:893]: This doesn't look right. **OTHER_FUNCTION** case should not be clubbed with any other function, as there can be more than one functions mapped to **OTHER_FUNCTION**.

File: core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java [hunk 2]
```diff
@@ -198,16 +198,22 @@ public HiveSqlDialect(Context context) {
...
 
   @Override public SqlOperator getTargetFunc(RexCall call) {
-    switch (call.type.getSqlTypeName()) {
-    case DATE:
-      switch (call.getOperands().get(1).getType().getSqlTypeName()) {
-      case INTERVAL_DAY:
-        if (call.op.kind == SqlKind.MINUS) {
-          return SqlLibraryOperators.DATE_SUB;
+    switch (call.getOperator().getKind()) {
+    case PLUS:
```
Comment (saranshmanglani) [inline core/src/main/java/org/apache/calcite/sql/dialect/HiveSqlDialect.java:202]: This part shouldn't be here. We already have a case check for **PLUS** and **MINUS** in the calling method.

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java [hunk 3]
```diff
@@ -120,10 +120,14 @@ public SnowflakeSqlDialect(Context context) {
   }
 
   @Override public SqlOperator getTargetFunc(RexCall call) {
-    switch (call.type.getSqlTypeName()) {
-    case DATE:
-    case TIMESTAMP:
-      return getTargetFunctionForDateOperations(call);
+    switch (call.getOperator().getKind()) {
```
Comment (saranshmanglani) [inline core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java:123]: Please remove all such **PLUS** and **MINUS** checks, as we already have a case check for the same in the calling method.

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java [hunk 4]
```diff
@@ -260,6 +271,19 @@ public SparkSqlDialect(SqlDialect.Context context) {
       default:
         return super.getTargetFunc(call);
       }
+    case OTHER_FUNCTION:
+      switch (call.type.getSqlTypeName()) {
```
Comment (saranshmanglani) [inline core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:275]: Please create a separate method for **OTHER_FUNCTION** handling.

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java [hunk 5]
```diff
@@ -644,6 +684,25 @@ private void unparseOtherFunction(SqlWriter writer, SqlCall call, int leftPrec,
       SqlWriter.Frame piFrame = writer.startFunCall("PI");
       writer.endFunCall(piFrame);
       break;
+    case "TO_HEX":
+      unparseUDF(writer, call, leftPrec, rightPrec, "UDF_CHAR2HEX");
```
Comment (saranshmanglani) [inline core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java:688]: Can we create an enum for the last parameter and then use it to get the desired value? 
That way we can club all the other such cases together and call this method only once.

File: core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java (hunk 1 above)
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java]: OTHER_FUNTION is no longer clubbed with PLUS and MINUS SqlKinds and getOperatorForOtherFunc is created and used instead.

File: core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java (hunk 3 above)
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/sql/dialect/SnowflakeSqlDialect.java]: done. removed.

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java (hunk 4 above)
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: done. created.

File: core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java (hunk 5 above)
Comment (Katakam-Phani-Mahidhar) [reply core/src/main/java/org/apache/calcite/sql/dialect/SparkSqlDialect.java]: done. created.

**Review by deepakkadam-dm — APPROVED** [review]
Reviewed by Saransh.

**Pull request merged**